### ISSUE TITLE
UrlStr gives unhelpful validation errors

### ISSUE THREAD

**Issue by cjw296** [samuelcolvin/pydantic#541]
# Bug

For bugs/questions:
* OS: macOS
* Python version `import sys; print(sys.version)`: 3.7.1
* Pydantic version `import pydantic; print(pydantic.VERSION)`: StrictVersion ('0.25')

Here's a short example:

```py
>>> from pydantic import BaseModel, urlstr
>>> class DatabaseConfig(BaseModel):
...     url: urlstr(schemes={'postgresql'}, require_tld=False) = ...
... 
>>> DatabaseConfig(url='postgresql://diary_test@localhost/diary_test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pydantic/main.py", line 235, in __init__
    object.__setattr__(self, '__values__', self._process_values(data))
  File "pydantic/main.py", line 438, in _process_values
    return validate_model(self, input_data)  # type: ignore
  File "pydantic/main.py", line 635, in validate_model
    raise ValidationError(errors)
pydantic.error_wrappers.ValidationError: 1 validation error
url
  url string does not match regex (type=value_error.url.regex)
```

So, couple of problems:

- This url is valid, it works ;-) 
- "does not match regex" isn't at all helpful to the end user, we should at least include the pattern being matched.

**Comment by samuelcolvin** [comment]
Problem is it's not that easy to work out what's wrong:

https://github.com/samuelcolvin/pydantic/blob/22f134475d8df84479239f37f68711fc569a05a9/pydantic/utils.py#L201-L228

**Comment by cjw296** [comment]
Yeah, that's not ideal ;-) Why not use something like urlparse that might give better feedback?
(and, as in this case, be correct...)

**Comment by samuelcolvin** [comment]
I guess because urlparse is either to lenient or too strict, but since we have the tests it shouldn't be too hard to try urlparse and see if you can get it to have match the current behavior + better error message and accepting the above.

**Comment by samuelcolvin** [comment]
as per the comment above, the code comes from marshmallow so you might find more about why that code was used in their issue tracker. Could be as simple as they (at some point) wanted to support an older version of python with a different or no `urlparse`.

**Comment by euri10** [comment]
Why not use DSN for db urls ?
```
>>> from pydantic import BaseModel
>>> from  pydantic import DSN, UrlStr
class C(BaseModel):
    d: DSN
    u: UrlStr
>>> C(d="postgresql://postgres:postgres@db:5432/postgres_test", u="postgresql://postgres:postgres@db:5432/postgres_test")
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/home/lotso/.local/share/virtualenvs/euri10_fastapi_base-9Y4Zbf8e/lib/python3.7/site-packages/pydantic/main.py", line 235, in __init__
    object.__setattr__(self, '__values__', self._process_values(data))
  File "/home/lotso/.local/share/virtualenvs/euri10_fastapi_base-9Y4Zbf8e/lib/python3.7/site-packages/pydantic/main.py", line 438, in _process_values
    return validate_model(self, input_data)  # type: ignore
  File "/home/lotso/.local/share/virtualenvs/euri10_fastapi_base-9Y4Zbf8e/lib/python3.7/site-packages/pydantic/main.py", line 635, in validate_model
    raise ValidationError(errors)
pydantic.error_wrappers.ValidationError: 1 validation error
u
  url scheme "postgresql" is not allowed (type=value_error.url.scheme; scheme=postgresql)
>>> C(d="postgresql://postgres:postgres@db:5432/postgres_test", u="http://test.com")
<C d='postgresql://postgres:postgres@db:5432/postgres_test' u='http://test.com'>
```

**Comment by samuelcolvin** [comment]
I guess really they're the same thing and should be unified, but for this case, yes, in this case that should work.

`urlstr` is used when you want to modify the constraints of `UrlStr`.

**Comment by cjw296** [comment]
Oh, didn't even realise what DSN was...

**Comment by euri10** [comment]
the code that validates the regexp is a hit or miss
```
regex = url_regex_generator(relative=cls.relative, require_tld=cls.require_tld)
        if not regex.match(value):
            raise errors.UrlRegexError()
```
afaik there's no way to tell where or why a regexp failed.

this said, should you want more info about it, one way I see could be the following:

1. in the `url_regex_generator` function you can use [named groups](https://github.com/euri10/pydantic/blob/b58b02cc8d4f9e12de97e071e6d750fefba021e3/pydantic/utils.py#L226-L243) as shown below in a branch that I just made (passes current tests fine) which help the comprehension a little bit:

```
    tweaked = r''.join(
        (
            r'^',
            r'(' if relative else r'',
            r'(?P<SCHEME>[a-z0-9\.\-\+]*)://',  # scheme is validated separately
            r'(?P<BASICAUTH>[^:@]+?:[^:@]*?@|)',  # basic auth
            r'(?P<DOMAIN>(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+',
            r'(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|',  # domain...
            r'(?P<LOCAL>localhost)|',  # localhost...
            (r'(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.?)|' if not require_tld else r''),  # allow dotless hostnames
            r'(?P<IPV4>\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|',  # ...or ipv4
            r'(?P<IPV6>\[[A-F0-9]*:[A-F0-9:]+\]))',  # ...or ipv6
            r'(?P<PORT>:\d+)?',  # optional port
            r')?' if relative else r'',  # host is optional, allow for relative URLs
            r'(?P<ENDURL>/?|[/?]\S+)$',
        )
    )
    return re.compile(tweaked, re.IGNORECASE)
```

2. what could be done after (but I haven't tested it and it would for sure slow things) is take each part of the tuple above inside ` r''.join()`, as a sub-regexp, then validate orderly the string against each subregexp not with a match but a find (do we find it or not?) then , if it fails report the error,  you'd know if it fails in SCHEME, BASICAUTH, etc, etc

**Comment by euri10** [comment]
Could also rework the entire regex like
```
^(?=(SCHEME))?(?=(SCHEME BASICAUTH))?(?=(SCHEME BASICAUTHDOMAIN))?.
```
etc, etc up to the ENDURL from above, with uppercase names replaced by their regex pattern value from above and therefore test each lookahead subregex incrementally over the last one. 
Will test something tomorrow :flushed:

**Comment by samuelcolvin** [comment]
This should be fixed on #755 where I've completely written url parsing for pydantic, one of the improvements should be more descriptive error messages although it's very difficult to give universally helpful error messages on something like this.

Please review #755 and let me know if it solves this problem and if not what more it could do.

**Comment by cjw296** [comment]
I chucked a couple of comments on #755, but this looks great! Thanks for working on it :-)
